### PR TITLE
[tool] Add pigeon support to update-dependency

### PR DESCRIPTION
**Pull request by stuartmorgan**
Adds special-cased support to `update-dependency` for `pigeon`, so that updating the `pigeon` dependency will also regenerate the pigeon files, so long as the package follows repository conventions of using `pigeons/*.dart` for the input, and has the output fully configured in the input file via `PigeonOptions`.

Since the testing for this ran into a somewhat long-standing problem with the mock process setup in the tool tests, where it was easy to accidentally return a mock process for the wrong variant of a call if there are multiple calls to the same command (e.g., `dart pub get` followed by `dart run pigeon` here), I finally fixed that as part of this PR: mock process results can now specify any number of arguments, and those must be the first arguments to the call when it tries to use the mock. This provides safety against tests accidentally providing the wrong mock result, without making it significantly more complex to use in tests. (An alternate version would use a Mockito-style matcher on the entire command+argument list to get the return object, instead of going in order through a list scoped only by executable, but flutter/flutter had a system like that and it was *very* hard to figure out why mocks weren't being returned when you thought they should be due to matcher mistakes.)

Existing tests have been updated, adding extra arguments when there's potential ambiguity.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: script/tool/test/update_dependency_command_test.dart [hunk 1]
```diff
@@ -339,5 +343,127 @@ dev_dependencies:
...
+          ProcessCall(
+            'dart',
+            const <String>['run', 'pigeon', '--input', 'pigeons/bar.dart'],
+            package.path,
+          ),
+        ]),
+      );
+    });
+
+    test('warns when regerating pigeon if there are no pigeon files', () async {
```
Comment (tarrinneal) [inline script/tool/test/update_dependency_command_test.dart:385]: typo: regenerating

**Review by tarrinneal — COMMENTED** [review]
Can this be run with a list of plugins? Or is it only useful for running on the entire directory?

**Review by tarrinneal — APPROVED** [review]

**Comment by stuartmorgan** [comment]
> Can this be run with a list of plugins? Or is it only useful for running on the entire directory?

Almost all the repo commands (including this one) can target specific packages with the `--packages` flag (or with some other things like targeting all changed packages, but that's unlikely to be as useful here).

So for instance you could `git grep swiftOut` on the repo to find that `file_selector_macos`, `path_provider_foundation`, and `shared_preferences_foundation` use Pigeon's Swift generator, and then run this command with `--pub-package=pigeon --packages=file_selector_macos,path_provider_foundation,shared_preferences_foundation` to update pigeon to the latest version in just those packages, all at once. Just to take a random example 😉

**Comment by tarrinneal** [comment]
I was just thinking about if there was potentially an unsupported feature, or a bug for certain features. Obviously, we would prefer not to have that at all though.

**Comment by tarrinneal** [comment]
> AuTO LABel Is ReMoveD FOr FlUttER/PaCKagEs, pr: 3640, due to - thE STatUS or cheCK SuIte MAC_Arm64 iOs_PLaTfOrM_tEstS_sharD_5 MaSteR hAS fAIlED. pLEASe FIX thE ISsueS IdentIFied (Or defLAKE) BEfORe Re-ApplyING tHiS labEl.

This is just flake, obviously

**Pull request merged**